File: leetcode/14D Algo/3. Sliding Window/3. Longest Substring Without Repeating Characters.cpp
#include<iostream>
#include<vector>
#include<queue>
#include<list>
#include<stack>
#include<algorithm>
#include<unordered_set> 
#include<unordered_map> 
using namespace std;

/*
Given a string s, find the length of the longest substring without repeating characters.

Input: s = "abcabcbb"
Output: 3
--------- abc

Input: s = "pwwkew"
Output: 3
--------- wke
*/

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        if(s.size() <= 1){
            return s.size();
        }
        unordered_map<char, int> M;
        int res = 0; int size = s.size();
        int j = 0;
        for(int i = 0; i < size; ++i){
            if(M.find(s[i]) != M.end()){
                j = max(M[s[i]] + 1, j); // 有可能重复的在前面
            }
            M[s[i]] = i;
            res = max(res, i - j + 1);
        }
        return res;
    }
};

File: leetcode/3. Longest Substring Without Repeating Characters.java
import java.util.HashMap;

/*
Input: s = "abcabcbb"
Output: 3
Explanation: The answer is "abc", with the length of 3.

Input: s = "pwwkew"
Output: 3, "kew"
*/


class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        if(n < 2) return n;
        HashMap<Character, Integer> M = new HashMap<>();
        int res = 0, l = 0, r = 0;
        while(r < n){          
            if(!M.containsKey(s.charAt(r)) || M.get(s.charAt(r)) < l){
                M.put(s.charAt(r), r);
            }
            else{
                l = M.get(s.charAt(r)) + 1;
                M.replace(s.charAt(r), r);
            }
            res = Math.max(r - l + 1, res);
            ++r;
        }
        return res;
    }
}